backend/src/services/package/PackageCostService.ts

```diff
@@ -0,0 +1,106 @@
+            if(!await this.db.packageExists(packageId)) {
+                throw new Error('404: Package not found');
+            }
+
+            if (!await S3.checkIfPackageExists(packageId)) {
+                throw new Error('404: Package not found');
+            }
+
+            // Retreive the ZIP info from S3
+            const zipBuffer = await S3.getFileByKey(packageId);
+            if(zipBuffer == null){
+                throw new Error('404: Package not found');
+            }
+
+            // Estimate the standalone cost of the package
+            this.standaloneCost =  Math.floor((zipBuffer.length *3) / 4); // Based on ratio of base64 encoding
+            return this.standaloneCost;
+        } catch (error) {
+            Logger.logError("Error fetching package cost: ", error);
+            throw error;
+        }
+    }
+
+    async getCost(packageId: string) {
+        try {
+            // Estimate the cost of the package with dependencies
+            // Setup the package URL
+            const packageUrl = await this.db.getPackageURL(packageId);
+            const urlHandler = new URLHandler(packageUrl);
+            await urlHandler.setRepoURL();
+
+            // Fetch the package.json file from the GitHub repository
+            const url = `https://api.github.com/repos/${urlHandler.getOwnerName()}/${urlHandler.getRepoName()}/contents/package.json`;
+            const response = await axios.get(url, {
+                headers: { 'Accept': 'application/vnd.github.v3.raw' }
+            });
+            const packageJson = response.data;
+
+            const dependencies = packageJson.dependencies || {};
+            const devDependencies = packageJson.devDependencies || {};
+
+            Logger.logDebug(`Dependencies: ${dependencies}`);
+            Logger.logDebug(`Dev Dependencies: ${devDependencies}`);
+
+            // Estimate the cost of the dependencies
+            let dependencyCost = 0;
+            for (const [packageName, packageVersion] of Object.entries(dependencies)) {
+                dependencyCost += await this.getDependencyCost(packageName, packageVersion as string);
+            }
+            for (const [packageName, packageVersion] of Object.entries(devDependencies)) {
+                dependencyCost += await this.getDependencyCost(packageName, packageVersion as string);
+            }
+
+            return this.standaloneCost + dependencyCost;
+        } catch (error) {
+            Logger.logError("Error fetching package cost: ", error);
+            throw error;
+        }
+    }
+
+    private cleanVersion(version: string): string{
+        return version.replace('^', '').replace('~', '');
+    }
+
+    async getDependencyCost(packageName: string, packageVersion: string) {
+        try {
+            Logger.logDebug(`Fetching dependency:  ${packageName}, ${packageVersion}`);
+
+            // Get the tar file from npm registry
+            packageVersion = this.cleanVersion(packageVersion);
```

Comment:
This could be a bit  more complex than just replacing ^ and ~. 

Since some dependencies might rely on the same package but with different versions, should we try to find a version that matches both their requirements, or would we download two different versions? For example, with ~, which only allows the last digit (patch updates) to change, if a one req was ~1.2.3 and another was ~1.2.5, we should try to estimate cost with only 1.2.5 or would we download both? We could just do both if that would keep it simpler, but I'm not sure (It would be more accurate to do only 1 I think and we wouldn't have to estimate for both versions).

Additionally, I believe according to piazza, they want you to keep going recursively into the dependencies. In this case you would also need to watch out for circular dependencies (I think). I think it would make sense to (1) try and find a list of all dependencies(watch out for circular dependency infinite loops), (2) choose the best combination of packageName/Ver, then (3) calculate the final total Cost and the costs for each dependency.

I believe we may need to also support the "-" range format for versions also.